library/psa_crypto.c

```diff
@@ -1184,17 +1247,51 @@ static psa_algorithm_t psa_key_policy_algorithm_intersection(
 static int psa_key_algorithm_permits( psa_algorithm_t policy_alg,
                                       psa_algorithm_t requested_alg )
 {
+    /* A requested algorithm cannot be a wildcard. */
+    if( PSA_ALG_IS_WILDCARD( requested_alg ) )
+        return( 0 );
     /* Common case: the policy only allows requested_alg. */
     if( requested_alg == policy_alg )
         return( 1 );
-    /* If policy_alg is a hash-and-sign with a wildcard for the hash,
-     * and requested_alg is the same hash-and-sign family with any hash,
-     * then requested_alg is compliant with policy_alg. */
-    if( PSA_ALG_IS_HASH_AND_SIGN( requested_alg ) &&
-        PSA_ALG_SIGN_GET_HASH( policy_alg ) == PSA_ALG_ANY_HASH )
-    {
-        return( ( policy_alg & ~PSA_ALG_HASH_MASK ) ==
-                ( requested_alg & ~PSA_ALG_HASH_MASK ) );
+    /* Special cases: the policy is an algorithm collection. */
+    if( PSA_ALG_IS_WILDCARD( policy_alg ) )
```

Comment:
Here, it does serve a purpose. Without this check, the current code would allow operations where policy_alg is set to a specific-length algorithm with a tag length lower than the requested length.

Reply:
Removed in favor of inlined specific checks.

Reply:
I'm primarily concerned about readability and robustness, with code size being only a minor concern. I'd be fine with a structure that makes it clear that the check is to avoid doing complex stuff (potentially error-prone or costly) in the wildcard case. I'm not so happy when it isn't obvious whether the wildcard check is essential to the logic or not, or when it isn't obvious that the check will still be right if a new form of wildcard is added.

I find the new version of the code more readable. Here I can easily follow the logic of each check without having to keep track of unrelated nested checks. Thanks for making the change.